Make slurm job canceling more robust

daniel-wer opened the pull request:
### Description:
- From time to time some slurm jobs keep running after killing voxelytics (and therefore invoking `inner_handle_kill` in `slurm.py`)
- I suspect this is either due to a race condition where jobs are in between pending and running when the function is called or because there is a moment during startup where the job doesn't react to the SIGINT signal
- To circumvent this, `scancel` is invoked again after a short grace period which definitely kills all given job ids, regardless of state

### Issues:
- fixes sporadic leftover slurm jobs

### Todos:
Make sure to delete unnecessary points or to check all before merging:
 - [ ] Updated Changelog
 - [ ] Updated Documentation
 - [ ] Added / Updated Tests
 - [ ] Considered adding this to the Examples

robert-oleynik reviewed (approved):
LGTM.

Note: Checking for jobs to kill may be cleaner, but this is a lot simpler (and should not return any errors)

daniel-wer commented:
I dove into this and fixed [#4035](https://github.com/scalableminds/voxelytics/issues/4035) while I was at it.

Turns out, when instantiating multiple cluster executors in the same process, the SIGINT/SIGTERM signal handler was overwritten multiple times which often caused the original signal handler of the main voxelytics process to no longer be called (i.e. if one of the executors no longer was referenced). To fix this and clean up the signal handling, I changed the signal handler overwriting/chaining to be implemented at a class level so it will only happen once and also won't be affected if previous executors are no longer referenced.

Regarding the sleep duration, I learned that it doesn't have to be <5s since the main voxelytics signal handler and the cluster executor signal handler are _chained_ and run one after the other (cluster executor first). So I opted to use the same timeout length that voxelytics is using (SIGTERM_WAIT_IN_S env variable - 5s by default).

philippotto commented on cluster_tools/cluster_tools/schedulers/cluster_executor.py:
```diff
@@ -38,18 +38,6 @@
 _S = TypeVar("_S")
 
 
-def _handle_kill_through_weakref(
```
I think the same function still exists in `dask.py`. is this alright?

daniel-wer replied:
@normanrz Do we still want to maintain the dask executor?

If so, it needs to be adapted as well, yes.

normanrz replied:
Nope. The dask executor can go.

daniel-wer replied:
We decided not to remove the dask executor at this point to avoid a breaking change. Not fixing this bug for it is not a big deal. We should deprecate the dask and kubernetes executors in a followup PR.

philippotto commented on cluster_tools/cluster_tools/schedulers/cluster_executor.py:
```diff
@@ -158,26 +141,68 @@ def as_completed(cls, futs: list[Future[_T]]) -> Iterator[Future[_T]]:
+            print(f"Error during shutdown: {e}")
+
+        if callable(existing_signal_handler) and existing_signal_handler not in (
+            signal.SIG_DFL,
+            signal.SIG_IGN,
+            signal.default_int_handler,
```
can you add a comment that explains this condition?

philippotto commented on cluster_tools/cluster_tools/schedulers/cluster_executor.py:
```diff
@@ -158,26 +141,68 @@ def as_completed(cls, futs: list[Future[_T]]) -> Iterator[Future[_T]]:
+
+    @classmethod
+    def _register_shutdown_hook(cls, hook: Callable[[], None]) -> None:
+        cls._shutdown_hooks.append(hook)
```
`_register_shutdown_hook` is only called in `__init__`. why is it a list then? couldnt it be an `Optional[Callable]` and `_installed_signal_handler` could also be absorbed, because either the hook was attached (is not none) or not.
additionally, I think that one could merge `_register_shutdown_hook` and `_install_signal_handlers` into one function, because one simply delegates to the other (plus appending to a list). I feel like this would be a bit easier to understand then.

philippotto reviewed (commented):
code looks good overall :+1: however, I still have some questions and also didn't test yet. maybe we can have a quick call (then you also don't need to reply to my questions here)?

> Turns out, when instantiating multiple cluster executors in the same process, the SIGINT/SIGTERM signal handler was overwritten multiple times which often caused the original signal handler of the main voxelytics process to no longer be called (i.e. if one of the executors no longer was referenced).

is this case tested? I don't see a spec which instantiates multiple executors 🤔 

> To fix this and clean up the signal handling, I changed the signal handler overwriting/chaining to be implemented at a class level so it will only happen once and also won't be affected if previous executors are no longer referenced.

I don't really understand this 🙈 questions for a call: why is a weakref no longer necessary? how is it ensured that it only happens once?

daniel-wer commented on cluster_tools/cluster_tools/schedulers/cluster_executor.py:
```diff
@@ -130,15 +121,7 @@ def __init__(
-            signal.SIGINT,
-            partial(_handle_kill_through_weakref, ref(self), existing_sigint_handler),
-        )
+        self._register_shutdown_hook(self.handle_kill)
```
Deregister in shutdown and test that afterwards this list was cleaned up

daniel-wer commented on cluster_tools/cluster_tools/schedulers/cluster_executor.py:
```diff
@@ -158,26 +141,68 @@ def as_completed(cls, futs: list[Future[_T]]) -> Iterator[Future[_T]]:
+    @classmethod
+    def _register_shutdown_hook(cls, hook: Callable[[], None]) -> None:
+        cls._shutdown_hooks.append(hook)
+        cls._install_signal_handlers()
```
Rename to ensure_signal_handler_is_installed or so

daniel-wer commented on cluster_tools/pyproject.toml:
```diff
@@ -17,15 +17,16 @@ Changelog = "https://github.com/scalableminds/webknossos-libs/blob/master/cluste
 [project.optional-dependencies]
-kubernetes = ["distributed ~=2023.9.1"]
-dask = ["kubernetes ~=27.2.0"]
+dask = ["distributed ~=2023.9.1"]
+kubernetes = ["kubernetes ~=27.2.0"]
```
The dependencies were switched by mistake :shrug:

philippotto commented on cluster_tools/tests/test_slurm.py:
```diff
@@ -168,85 +169,149 @@ def wait_until_first_job_was_submitted(
+    )
+
+    assert len(cluster_tools.SlurmExecutor._shutdown_hooks) == 0
+    # Reset class variable after changing signal handler so it will be set again when the next cluster executor is instantiated
+    cluster_tools.SlurmExecutor._installed_signal_handler = False
```
won't this lead to the signal handler being set up twice? how about a class method à la `uninstall_signal_handler` which would do the uninstall + set the variable to false? however, it might be overkill if it's only necessary for this test.

daniel-wer replied:
In lines 262-264 we "reset" the signal handler, so the cluster executor signal handler that was installed by one of the previous tests is no longer installed. Therefore, we also need to make sure the `_installed_signal_handler` class variable is set to False again.
It's a special case for this test, because we pretend that a signal handler is already installed before the cluster executor is instantiated (to mimic what happens if voxelytics instantiates a cluster executor) and check whether it was actually called, then.

philippotto reviewed (commented):
code looks good :+1: thank you so much for spending your time to get this right!
I didn't approve though, because you mentioned long delays during testing..
